File: competitive-programming/leet-code/207/canFinish.js
var canFinish = function (numCourses, prerequisites) {
    const adj = Array.from({ length: numCourses }, () => []);
    for (const [course, pre] of prerequisites) {
        adj[pre].push(course);
    }

    const vis = new Array(numCourses).fill(false);
    const path = new Array(numCourses).fill(false);

    const dfs = (node) => {
        vis[node] = path[node] = true;

        for (const next of adj[node]) {
            if (!vis[next]) {
                if (dfs(next)) return true;
            } else if (path[next]) {
                return true;
            }
        }

        path[node] = false;
        return false;
    };

    for (let i = 0; i < numCourses; i++) {
        if (!vis[i]) {
            if (dfs(i)) return false;
        }
    }

    return true;
};

File: competitive-programming/leet-code/207/can_finish.py
from typing import List

def can_finish(num_courses: int, prerequisites: List[List[int]]) -> bool:
    adj = [[] for _ in range(num_courses)]
    for course, pre in prerequisites:
        adj[pre].append(course)

    visited = [False] * num_courses
    in_path = [False] * num_courses

    def dfs(node: int) -> bool:
        visited[node] = True
        in_path[node] = True

        for next_course in adj[node]:
            if not visited[next_course]:
                if dfs(next_course):
                    return True
            elif in_path[next_course]:
                return True

        in_path[node] = False
        return False

    for i in range(num_courses):
        if not visited[i]:
            if dfs(i):
                return False

    return True